token/program-2022/src/offchain.rs

```diff
@@ -46,14 +61,1005 @@ where
+/// use solana_program::pubkey::Pubkey;
+///
+/// let mint = Pubkey::new_unique();
+/// let client = RpcClient::new_mock("succeeds".to_string());
+/// let mut account_metas = vec![];
+///
+/// get_extra_transfer_account_metas(
+///     &mut account_metas,
+///     |address| self.client.get_account(&address).map_ok(|opt| opt.map(|acc| acc.data)),
+///     &mint,
+/// ).await?;
+/// ```
+/// Note that this offchain helper will build a new `Execute` instruction,
+/// resolve the extra account metas, and then add them to the transfer
+/// instruction. This is because the extra account metas are configured
+/// specifically for the `Execute` instruction, which requires five accounts
+/// (source, mint, destination, authority, and validation state), wheras the
+/// transfer instruction only requires four (source, mint, destination, and
+/// authority) in addition to `n` number of multisig authorities.
+#[allow(clippy::too_many_arguments)]
+pub async fn create_transfer_instruction_with_extra_metas<F, Fut>(
+    token_program_id: &Pubkey,
+    source_pubkey: &Pubkey,
+    mint_pubkey: &Pubkey,
+    destination_pubkey: &Pubkey,
+    authority_pubkey: &Pubkey,
+    signer_pubkeys: &[&Pubkey],
+    amount: u64,
+    decimals: u8,
+    fetch_account_data_fn: F,
+) -> Result<Instruction, AccountFetchError>
+where
+    F: Fn(Pubkey) -> Fut,
+    Fut: Future<Output = AccountDataResult>,
+{
+    let mut transfer_ix = crate::instruction::transfer_checked(
+        token_program_id,
+        source_pubkey,
+        mint_pubkey,
+        destination_pubkey,
+        authority_pubkey,
+        signer_pubkeys,
+        amount,
+        decimals,
+    )?;
+
+    if token_program_id == &crate::id() {
+        let mint_data = fetch_account_data_fn(*mint_pubkey)
+            .await?
+            .ok_or(ProgramError::InvalidAccountData)?;
+        let mint = StateWithExtensions::<Mint>::unpack(&mint_data)?;
+
+        if let Some(program_id) = transfer_hook::get_program_id(&mint) {
+            // Convert the transfer instruction into an `Execute` instruction,
+            // then resolve the extra account metas as configured in the validation
+            // account data, then finally add the extra account metas to the original
+            // transfer instruction.
+            let mut execute_ix = spl_transfer_hook_interface::instruction::execute(
+                &program_id,
+                source_pubkey,
+                mint_pubkey,
+                destination_pubkey,
+                authority_pubkey,
+                &get_extra_account_metas_address(mint_pubkey, &program_id),
+                amount,
+            );
+
+            resolve_extra_account_metas(
+                &mut execute_ix,
+                fetch_account_data_fn,
+                mint_pubkey,
+                &program_id,
+            )
+            .await?;
+
+            transfer_ix
+                .accounts
+                .extend_from_slice(&execute_ix.accounts[5..]);
```

Comment:
This seems a bit confusing and error-prone, and shows that `resolve_extra_account_metas` is also a problem, since it's easy to misuse if you don't provide an `execute` instruction.

We should also deprecate `resolve_extra_account_metas` and add a new helper function in the interface that takes in the params and a mutable instruction. It adds the required accounts there, similar to what's happening here. We can call it something like `add_extra_account_metas_for_execute` or something.

What do you think?

Reply:
I think this makes sense. In theory, if we wanted to provide a helper that can add extra metas to _any_ instruction (which we know now is... a bit sketchy), it should probably live in `tlv-account-resolution`.

Since this is specifically a `spl-transfer-hook-interface` helper, yeah I agree it should be robust and designed specifically to create `execute` instructions.